### PR TITLE
Add modding API and refactor the CSM project

### LINKED ISSUE #204: TM:PE seems to be incompatible with NLog
Simply launching TMPE and CSM together results in this error:

image

`Warning 15.9578899: Could not load global config: System.TypeLoadException: Could not load type 'System.Xml.Serialization.XmlRootAttribute' from assembly 'System.Xml, Version=2.0.0.0, ...`

I am using [the latest CI version](https://ci.appveyor.com/project/krzychu124/tmpe/branch/master/artifacts), since https://github.com/CitiesSkylinesMods/TMPE/pull/866 is not contained in any releases.

I have no idea why this is happening, but removing NLog fixes the problem.

### LINKED ISSUE #264: Airport Updates
**Describe the bug**
Sync problems, random holes in the ground, zoning and building not appearing
**To Reproduce**

**Expected behavior**

**Screenshots**

**Active DLCs**
Please add a list of your enabled DLCs here.

**Active mods**
Please add a list of your installed and enabled mods here.

**Versions**
Type /version into the in-game chat and paste the version numbers here:
- Mod version: 
- Game version: 

**Additional context**

### LINKED ISSUE #265: Airport DLC issues
**Describe the bug**
You cannot place a road without a terrain bug. We try to de-activate all workshop mods regarding road that can maybe create the problem, play a different map, change the host... it is broken by the DLC.

**To Reproduce**
- Create a multiplayer
- Join in
- Place road... bug occurs

**Screenshots**
Capture

**Active DLCs**
All DLCs, except Natural Disasters

**Active mods**
Fine Road Tool
Auto Bulldoze
Precision Engineering
customize Bus livery (TPG addon)

**Versions**
Type /version into the in-game chat and paste the version numbers here:
- Game version: last version released with Airport DLC on January 25 2022.

Hope you will solve this bug, because your MOD is AMAZING!

### LINKED ISSUE #268: Can't use multiplayer menu when using 4k resolution  in fullscreen
**Describe the bug**
In 4k fullscreen the multiplayer menu is half hidden in the bottom right corner making it impossible to interact with any option. See screenshot. Changing the resolution to HD Windowed and then restarting the game fixes the bug (but its super annoying because if I change the resolution back to 4k the bug is back and I would have to restart the game to fix it again)

**To Reproduce**
1. Set the resolution to 4k in fullscreen

*Expected behavior**
Menu is actually accessible.

**Screenshots**

**Active DLCs**
Snowfall
After Dark
Mass Transit

**Active mods**
none

**Versions**
Type /version into the in-game chat and paste the version numbers here:
- Mod version: 2201. 180.0.0
- Game version: 1.14.0-f9

**Additional context**
Using Windows 10 with NVIDIA RTX 2080 GPU and playing (as you guessed) on a 4K display in 60 Hz.

### PR DESCRIPTION
**Pull request by kaenganxt**
This patch includes multiple features and changes, that will be explained in the following:

### Modding API
- The new API allows to:
  - Create own commands and handlers
  - Send commands
  - Print messages in chat
  - Use the IgnoreHelper or ReflectionHelper classes
  - Log messages (Err/Warn/Info/Debug)
- Mods require extending a Connection class with the following functionality:
  - Fields to set: A name, an enabled flag (can be used if the mod support is part of our project and the class has to detect if the mod is there), the mod class (used for checking compatibility)
  - A list of assemblies to check for commands: The protocol is automatically adapted depending on the enabled mod support classes.
  - RegisterHandlers and UnregisterHandlers methods for (un)registering logic handlers (like for example Harmony patches)
- Add example mod that uses the new modding api in `examples/SampleExternalMod`
- Require equal mods on both sides when connecting
  - Only consider mods with registered mod support for now (Others may include client side mods that we want to ignore)
  - Additionally consider Cities' built-in mods
  - Also require installed assets to be the same
- Display mod support status when starting a server/joining a game with the following states:
  - Supported: Either a mod support class is registered or the mod is explicitly supported by CSM (e.g. the remove chirper mod)
  - Known working: Matched against a list of mods known to work (will be expanded in the future)
  - Unsupported: Matched against a list of mods that don't work (e.g. TM:PE for now)
  - Unknown: None of the other options matched
  - Client side mod: Additional state in parallel to one of the others, matched against a list of mods

### Refactor
- The project now consists of multiple subprojects:
  - CSM: The core, including the mod class and networking logic
  - CSM.API: The modding API, mods can be compiled against this project. It is also uploaded to Nuget as CitiesSkylinesMultiplayer.API (CSM was a reserved namespace) https://www.nuget.org/packages/CitiesSkylinesMultiplayer.API/
  - CSM.BaseGame: The base game syncing logic, this module is registered as a mod to the CSM core and thus only needs the API to function

- Instead of bundling Harmony with this project, now the [CitiesHarmony mod](https://github.com/boformer/CitiesHarmony) is required as a dependency. This was done to be better compatible to other mods that do the same. When enabling CSM and CitiesHarmony is not found, the user is prompted to install it either automatically (on Steam) or manually (on other platforms). Additionally, CitiesHarmony can be added as a dependency in the Steam Workshop so that the user is already prompted to install it before starting the game.

- Replace NLog with own logging mechanism as it caused problems with some other mods. (fixes #204)

### New features (From the Airports update)
- Sync changed service/public transport vehicles
- Sync soil trade
- Sync street tree upgrades (Changing trees on streets)
- As all base game features of the Airport update are now implemented: Closes #265, closes #264

### Fixed issues
- Fix crash with Loading Screen Mod, we now start loading in the correct thread
- Center windows correctly (also works with the More Aspect Ratios mod) (fixes #268)

### TODO after merge
- Add CitiesHarmony dependency to Steam Workshop item

Thanks to @joaogl and @MightyWizard for starting with this modding API.

### CODE REVIEW

**Comment by DominicMaas** [comment]
This is mighty impressive!!

**Review by DominicMaas — COMMENTED** [review]
I'd imagine that we need to upgrade the build script to take into account VS2022. I don't mind doing this once I'm home as I'm currently running 2022.

File: src/csm/Networking/Status/ClientStatus.cs [hunk 1]
```diff
@@ -1,4 +1,5 @@
-﻿namespace CSM.Networking.Status
+﻿
```
Comment (DominicMaas) [inline src/csm/Networking/Status/ClientStatus.cs]: whitespace

**Comment by DominicMaas** [comment]
From the code side of things, it looks good. Love the idea of having the base game functionality as a mod itself. Helps validate that the mod support / apis are actually useful. 

Also agree with the project restructure here, will help with things going forward in the future.

**Comment by DominicMaas** [comment]
From what I understand, mod authors must bundle in the API dll when they distribute their project? Or do they just build against it, and then use the API dll that's included with CSM (since there's no point installing a multiplayer mod, without csm)?

**Comment by kaenganxt** [comment]
Yes they need to bundle the API dll. Sadly Cities fails to load an Assembly even if a single class could not be loaded due to a missing type. I already updated the wiki page to also include this fact: https://github.com/CitiesSkylinesMultiplayer/CSM/wiki/modding-api#distributing-your-mod-with-csm-support

**Comment by DominicMaas** [comment]
Ah that's annoying, We probably won't want to update the API dll often at all, I'd imagine we would encounter some type of version hell when updating CSM, and then the mods (with different API versions during the overlap period)

**Comment by DominicMaas** [comment]
Also, do you know if CitiesHarmony cause conflicts with CSM at the moment? Or are they fine side-by-side (should I update the steam listing now with the dependency, or wait until this version has been released?)

**Comment by kaenganxt** [comment]
I'm not sure if there would be conflicts but I guess it's best to wait for the release of this version.

**Comment by kaenganxt** [comment]
I guess this is ready from my side now, the modding API is fully documented and I'll update the further documentation once we release the update.

**Comment by DominicMaas** [comment]
Awesome, code looks good here. Lets get this merged.

**Review by DominicMaas — APPROVED** [review]

**Pull request merged**